### PR TITLE
Positioning accuracy indicator and digitizing accuracy requirement

### PR DESCRIPTION
**Pull request by nirvn**
This PR adds a new accuracy indicator features as well as an accuracy requirement enforcement mechanism when digitizing features. 

When the accuracy indicator is enabled, a badge is attached to the location button and colored red if the accuracy value is below bad, yellow if it falls short of excellent, or green. The badge is subtle as to avoid obscuring the map area:
image

In addition, an accuracy requirement mode can be toggled on alongside the indicator. The mode is active when the coordinate cursor to be locked to the GPS/NMEA position and prohibits vertex addition when the position accuracy value is below the bad accuracy value threshold. A toaster message is displayed for the user to understand he/she needs to wait for the device to gain a better accuracy:
image

All of the above can be  toggled and tweaked in the QField settings' positioning panel:
image

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for c7cf6b9bfbb5e8a48defe32d8e07b366a2734327:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-207e6ec5dc0384a16dc0c2d2aa068d98b374baae-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-207e6ec5dc0384a16dc0c2d2aa068d98b374baae-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-207e6ec5dc0384a16dc0c2d2aa068d98b374baae-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-207e6ec5dc0384a16dc0c2d2aa068d98b374baae-x86.apk)

File: src/qml/qgismobileapp.qml [hunk 1]
```diff
@@ -1050,6 +1057,22 @@ ApplicationWindow {
...
+              rightMargin: 2
+              topMargin: 2
+          }
+
+          width: 9
+          height: 9
+          radius: width / 2
+
+          visible: positioningSettings.accuracyIndicator && gpsButton.state === "On"
+          color: !positionSource.positionInfo || !positionSource.positionInfo.haccValid || positionSource.positionInfo.hacc > positioningSettings.accuracyBad ? "#e41a1c" : positionSource.positionInfo.hacc > positioningSettings.accuracyExcellent ? "#ff7f00" : "#4daf4a"
```
Comment (m-kuhn) [inline src/qml/qgismobileapp.qml:1074]: Shall we put these values into the theme config? Or some other config place?

File: src/qml/QFieldSettings.qml [hunk 2]
```diff
@@ -361,6 +361,173 @@ Page {
...
+                          if( text.length === 0 || isNaN(text) ) {
+                              positioningSettings.accuracyBad = NaN
+                          } else {
+                              positioningSettings.accuracyBad = parseFloat( text )
+                          }
+                      }
+                  }
+
+                  Label {
+                      text: qsTr("Excellent accuracy above [m]")
```
Comment (m-kuhn) [inline src/qml/QFieldSettings.qml:450]:
```suggestion
                      text: qsTr("Excellent accuracy %1 above [m]").arg('<font color="#e41a1c">•</font>')
```

What do you think about showing the color here?
Green for excellent
Yellow for acceptable

Comment (nirvn) [reply src/qml/QFieldSettings.qml:450]: Not a bad idea. Maybe we can show those in the gray explanatory paragraphs instead.

File: src/qml/qgismobileapp.qml (hunk 1 above)
Comment (nirvn) [reply src/qml/qgismobileapp.qml:1074]: Actually I think we already have all the colors we need in theme, my bad.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for c1b5914671c2a98a967bc3fb6c4cd65f1163c295:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-46a29a33200a3a5b62e0c5ec54fb5de31202423f-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-46a29a33200a3a5b62e0c5ec54fb5de31202423f-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-46a29a33200a3a5b62e0c5ec54fb5de31202423f-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-46a29a33200a3a5b62e0c5ec54fb5de31202423f-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 0c1f3f76867b83b37d4a91ceaefdf178a6973667:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-0bdb30ead67240eee03a978fcafd2f063c8db333-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-0bdb30ead67240eee03a978fcafd2f063c8db333-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-0bdb30ead67240eee03a978fcafd2f063c8db333-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-0bdb30ead67240eee03a978fcafd2f063c8db333-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 5b1e0db67245b05336de82a99dd8a37c7d5a2bc0:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-923c6a100759f67a0be9f64ad568c23c0079263c-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-923c6a100759f67a0be9f64ad568c23c0079263c-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-923c6a100759f67a0be9f64ad568c23c0079263c-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-923c6a100759f67a0be9f64ad568c23c0079263c-x86.apk)

File: src/qml/imports/Theme/Theme.qml [hunk 3]
```diff
@@ -63,5 +63,10 @@ QtObject {
       var path = '/themes/' + theme + '/nodpi/' + name + '.svg';
       return path;
     }
+
+    function colorToHtml(color) {
+      return "rgba(%1,%2,%3,%4)".arg(Math.floor(Theme.errorColor.r * 255)).arg(Math.floor(Theme.errorColor.g * 255)).arg(Math.floor(Theme.errorColor.b * 255)).arg(Math.floor(Theme.errorColor.a * 255));
```
Comment (suricactus) [inline src/qml/imports/Theme/Theme.qml:68]:
```suggestion
      return "rgba(%1,%2,%3,%4)"
      	.arg(Math.floor(Theme.errorColor.r * 255))
      	.arg(Math.floor(Theme.errorColor.g * 255))
      	.arg(Math.floor(Theme.errorColor.b * 255))
      	.arg(Math.floor(Theme.errorColor.a * 255));
```

File: src/qml/qgismobileapp.qml [hunk 4]
```diff
@@ -1050,6 +1057,22 @@ ApplicationWindow {
...
+              rightMargin: 2
+              topMargin: 2
+          }
+
+          width: 9
+          height: 9
+          radius: width / 2
+
+          visible: positioningSettings.accuracyIndicator && gpsButton.state === "On"
+          color: !positionSource.positionInfo || !positionSource.positionInfo.haccValid || positionSource.positionInfo.hacc > positioningSettings.accuracyBad ? Theme.errorColor : positionSource.positionInfo.hacc > positioningSettings.accuracyExcellent ? Theme.warningColor : Theme.mainColor
```
Comment (suricactus) [inline src/qml/qgismobileapp.qml:1074]:
```suggestion
          color: !positionSource.positionInfo
          	|| !positionSource.positionInfo.haccValid
          	|| positionSource.positionInfo.hacc > positioningSettings.accuracyBad 
          		? Theme.errorColor 
          		: positionSource.positionInfo.hacc > positioningSettings.accuracyExcellent 
          			? Theme.warningColor 
          			: Theme.mainColor
```

File: src/qml/qgismobileapp.qml [hunk 5]
```diff
@@ -465,6 +465,13 @@ ApplicationWindow {
       mapSettings: mapCanvas.mapSettings
       currentLayer: dashBoard.currentLayer
       overrideLocation: gpsLinkButton.linkActive ? positionSource.projectedPosition : undefined
+      accuracyRequirementFail: {
+          if ( positioningSettings.accuracyIndicator && positioningSettings.accuracyRequirement )
+          {
+              return positioningSettings.accuracyBad > 0 && ( !positionSource.positionInfo || !positionSource.positionInfo.haccValid || !positionSource.positionInfo.hacc > badThreshold )
```
Comment (suricactus) [inline src/qml/qgismobileapp.qml:471]:
```suggestion
              return positioningSettings.accuracyBad > 0 
              	&& ( 
              		!positionSource.positionInfo
              		|| !positionSource.positionInfo.haccValid 
              		|| !positionSource.positionInfo.hacc > badThreshold
                )
```
is the last condition ok with the `!` in front?

Comment (m-kuhn) [reply src/qml/qgismobileapp.qml:471]: That looks weird. It should just be without the `!` (i.e. if current accuracy is ">" == "less precise" than the threshold)

File: src/qml/imports/Theme/Theme.qml [hunk 6]
```diff
@@ -63,5 +63,10 @@ QtObject {
       var path = '/themes/' + theme + '/nodpi/' + name + '.svg';
       return path;
     }
+
```
Comment (suricactus) [inline src/qml/imports/Theme/Theme.qml:66]:
```suggestion

function toInlineStyles(styleProperties) {
  var styles = ''
  
  for (var property in styleProperties) {
    var value = styleProperties[property];
    style += property
    style += ': '
    style += typeof value == 'color'
      ? colorToHtml(value)
      : value
    style += ';'
  }
  
  return styles;
}

```
Have not tested, but I would suggest setting styles like:
```
  'String with <span style="%1">styles</a>'.arg(toInlineStyles({
    color: Theme.mainColor,
    weight: checkboxComponent.checked ? 'bold' : 'normal'
  }))
```

Comment (m-kuhn) [reply src/qml/imports/Theme/Theme.qml:66]: Sounds good.
Can you push this approach directly to this branch?

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 74102990c91b6661a9517fcbe269d039a70176f9:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-2498ecdd51fd7939495ad596f53482764f1c11c0-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-2498ecdd51fd7939495ad596f53482764f1c11c0-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-2498ecdd51fd7939495ad596f53482764f1c11c0-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-2498ecdd51fd7939495ad596f53482764f1c11c0-x86.apk)

**Comment by nirvn** [comment]
(About to push @suricactus 's suggestions)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for e0cd54f6243200107bbc3db74c62bb43e7743008:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-c9fd7acc640c06b06d65de2b5ba551c4a6762467-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-c9fd7acc640c06b06d65de2b5ba551c4a6762467-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-c9fd7acc640c06b06d65de2b5ba551c4a6762467-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1556-c9fd7acc640c06b06d65de2b5ba551c4a6762467-x86.apk)

**Review by m-kuhn — APPROVED** [review]
Great work!

**Pull request merged**